src/ast.ts

```diff
@@ -0,0 +1,63 @@
+export interface Syn {
+    readonly type: string;
+    readonly range?: [number, number];
+    readonly loc?: SourceLocation;
+}
+
+export interface Position {
+    readonly line: number;
+    readonly column: number;
+}
+
+export interface SourceLocation {
+    readonly start: Position;
+    readonly end: Position;
+    readonly source?: string | null;
```

Comment:
does this need to be string | null if it's already optional?

Reply:
The ESprima spec, which I generally follow, does use the optional `string | null`, but I don't see a reason not to change this to a mandatory `string | null`.